src/test/java/net/minecraftforge/test/GuiTabsTest.java

```diff
@@ -0,0 +1,243 @@
+                @Override
+                public void onTabClicked(GuiContainer guiContainer)
+                {
+                    super.onTabClicked(guiContainer);
+                    GuiTabsTest.INSTANCE.sendToServer(new TabMessage(3));
+                }
+            }.addTo(GuiInventory.class));
+            IntStream.range(0, 17).forEach(i -> new GuiTab("Tab-2" + i, new ItemStack(Items.DIAMOND), GuiBeacon.class)
+            {
+                @Override
+                public void onTabClicked(GuiContainer guiContainer)
+                {
+                    super.onTabClicked(guiContainer);
+                    GuiTabsTest.INSTANCE.sendToServer(new TabMessage(4));
+                }
+            }.addTo(GuiInventory.class));
+        }
+
+    }
+
+    public static class TestGui extends GuiChest
+    {
+        public TestGui(IInventory upperInv, IInventory lowerInv)
+        {
+            super(upperInv, lowerInv);
+        }
+    }
+
+    public static class TestGui2 extends TestGui
+    {
+        public TestGui2(IInventory upperInv, IInventory lowerInv)
+        {
+            super(upperInv, lowerInv);
+        }
+    }
+
+    public static class GuiHandler implements IGuiHandler
+    {
+
+        IInventory invC;
+        IInventory invS;
+
+        @Override
+        public Object getServerGuiElement(int ID, EntityPlayer player, World world, int x, int y, int z)
+        {
+            if (invS == null)
+            {
+                invS = new InventoryBasic("Test", true, 27);
+            }
+            return new ContainerChest(player.inventory, invS, player);
+        }
+
+        @Override
+        public Object getClientGuiElement(int ID, EntityPlayer player, World world, int x, int y, int z)
+        {
+            if (invC == null)
+                invC = new InventoryBasic("Test", true, 27);
+            switch (ID)
+            {
+            case 0:
+                return new TestGui(player.inventory, invC);
+            case 1:
+                return new GuiChest(player.inventory, invC);
+            case 2:
+                return new TestGui2(player.inventory, invC);
+            case 3:
+                return new GuiChest(player.inventory, invC);
+            case 4:
+                return new GuiBeacon(player.inventory, invC);
+            }
+            return null;
+        }
+
+    }
+
+    public static class TabMessageHandler implements IMessageHandler<TabMessageHandler.TabMessage, IMessage>
+    {
+
+        @Override
+        public IMessage onMessage(TabMessageHandler.TabMessage message, MessageContext ctx)
+        {
+            if (!ctx.getServerHandler().player.world.isRemote)
+            {
+                ctx.getServerHandler().player.openGui(GuiTabsTest.instance, message.id, ctx.getServerHandler().player.world,
+                        ctx.getServerHandler().player.chunkCoordX, ctx.getServerHandler().player.chunkCoordY,
+                        ctx.getServerHandler().player.chunkCoordZ);
```

Comment:
these are the player's chunk coords (world coords / 16). is there any reason they're being used? if the gui handler doesn't need them just pass zeroes instead, it's cleaner.

Reply:
Ah okay, until today I thought they were like a requirement, (I know its stupid but I thought it might need the exact location of the player to open the gui). replaced with 0s